DEPRECATION.md

```diff
@@ -19,55 +26,80 @@ underscore (`_`).
+Similarly, while it is permissible where necessary for behavior to change with no single-code path to support both the last minor of one major release and the first minor of a new major release, it is still strongly preferable if you can achieve this.
+
+
+## What is the public interface?
 
-- never assume that a function that isn't explicitly internal isn't in use;
+> [!NOTE]
+> This section should be in sync with [the release schedule documentation of Qiskit](https://docs.quantum.ibm.com/start/install#release-schedule).
+> Please [open an issue against Qiskit](https://github.com/Qiskit/qiskit/issues/new/choose) if there are discrepancies so we can clarify them.
 
-- all deprecations, changes and removals are considered API changes, and can
-  only occur in minor releases not patch releases, per the [stable branch policy](https://github.com/Qiskit/qiskit/blob/main/MAINTAINING.md#stable-branch-policy).
+For the purposes of semantic versioning, the Qiskit public API comprises all *publicly documented* packages, modules, classes, functions, methods, and attributes.
+
+An object is *publicly documented* if and only if it appears in [the hosted API documentation](https://docs.quantum.ibm.com/api/qiskit) for Qiskit.
+The presence of a docstring in the Python source (or a `__doc__` attribute) is not sufficient to make an object publicly documented; this documentation must also be rendered in the public API documentation.
+
+As well as the objects themselves needing to be publicly documented, the only public-API *import locations* for a given object is the location it is documented at in [the public API documentation](https://docs.quantum.ibm.com/api/qiskit), and parent modules or packages that re-export the object (if any).
+For example, while it is possible to import `Measure` from `qiskit.circuit.measure`, this is not a supported part of the public API for two reasons:
+
+1. The module `qiskit.circuit.measure` is not publicly documented, so is not part of the public interface.
+2. The [`Measure` object is documented as being in `qiskit.circuit.library`](https://docs.quantum.ibm.com/api/qiskit/qiskit.circuit.library.Measure), and is re-exported by `qiskit.circuit`, so the public import paths are `from qiskit.circuit.library import Measure` and `from qiskit.circuit import Measure`.
+
+As a rule of thumb, if you are using Qiskit, you should import objects from the highest-level package that exports that object.
+
+From within the Qiskit package itself, objects should be imported from the highest-level package that does not cause cyclic-import concerns.
+In general, imports should not reach into the internals of *other* subpackages; `qiskit.circuit` should import all its objects from `qiskit.quantum_info` and not from submodules of that, though in practice, our package hierarchy is sufficiently confused that this is not always possible).
+Using package-level imports helps indicate when new code is producing confused inter-package dependencies.
```

Comment:
Is this necessary, my personal practice is normally to go as deep in the import "tree" as possible partially to avoid potential cycles, but also just out of habit.

Reply:
Personally I'd say yes it's necessary - imo, going deeper than is necessary is what exacerbates the package-hierarchy problems we have; it can sometimes make imports work that _shouldn't_ because they're topologically out-of-order and then local imports added later can't resolve the cyclic problems caused that _should_ work.

That said, that line is definitely me legislating style by fiat right now, and it would be better to discuss that.

Reply:
I've removed that little bit 8c6c560c8f - it's something we should talk about separately anyway, you're right.